# Project 4 -- [Niharika] [Rai]

**TA Help:** Sharan Sivakumar. (Question 2a). 
    
**Collaboration:** N/A

## Question 1

In [2]:
import pandas as pd
myDF = pd.read_csv("/anvil/projects/tdm/data/noaa/1880.csv" , header=None , names=["id","date","element_code","value","mflag","qflag","sflag","obstime"]) 
myDF.head() 

,id,date,element_code,value,mflag,qflag,sflag,obstime
0,AGE00147705,18800101,PRCP,0,NaN,NaN,E,NaN
1,AGE00147708,18800101,PRCP,36,NaN,NaN,E,NaN
2,AGE00147709,18800101,PRCP,36,NaN,NaN,E,NaN
3,AGE00147712,18800101,PRCP,0,NaN,NaN,E,NaN
4,AGE00147713,18800101,TMAX,52,NaN,NaN,E,NaN


In [4]:
for index, row in myDF.iterrows(): 
    if row["element_code"] == "PRCP":
        if row["value"] > 1200:
            print(row["date"]) 

18800108
18800109
18800316
18800420
18800706
18800805
18800806
18800806
18800811
18800827


In [6]:
myDF[(myDF.element_code == "PRCP") & (myDF.value > 1200)] ["date"] 

6968      18800108
8868      18800109
74914     18800316
109082    18800420
187316    18800706
218130    18800805
219182    18800806
219199    18800806
223669    18800811
241073    18800827
Name: date, dtype: int64

For part a and b, I got the code structure by watching the project 4 videos: https://the-examples-book.com/programming-languages/python/some-examples-for-TDM-10200-project-4. More specifically, I watched the first video, "Comparing iterrows and indexing" to get reference on how to formulate my code. For part c, the indexing method is faster than iterrows. Personally, I understand the indexing method more because it is a one line code, and all the requirements such as the PRCP and the value being greater than 1200 is written in a way that one would expect to write in Math classes. Since, I understand Math, I find this code structure to be more efficient. 

## Question 2

In [1]:
import pandas as pd
from pathlib import Path 

myDF = []
for year in range (1800, 1851):
    myDF = pd.read_csv(Path(f'/anvil/projects/tdm/data/noaa/{year}.csv'), header=None, names=["id","date","element_code","value","mflag","qflag","sflag","obstime"]) 
    prcp_data = myDF[myDF['element_code'] == 'PRCP']
    avg_precip = prcp_data['value'].mean() 
    print(f"{year}: {avg_precip}") 

1800: 16.016438356164382
1801: 19.416438356164385
1802: 11.37945205479452
1803: 19.63013698630137
1804: 16.968903436988544
1805: 13.963013698630137
1806: 13.481361426256077
1807: 15.396825396825397
1808: 16.23633879781421
1809: 14.687671232876712
1810: 14.23489010989011
1811: 12.754794520547945
1812: 15.904371584699453
1813: 22.049397590361444
1814: 21.04501216545012
1815: 24.774074074074075
1816: 20.02589395807645
1817: 20.357409713574096
1818: 18.72570725707257
1819: 23.2260101010101
1820: 17.293478260869566
1821: 19.30807453416149
1822: 15.856035437430787
1823: 18.665745856353592
1824: 16.27336860670194
1825: 16.06637168141593
1826: 18.396419437340153
1827: 17.94869053981828
1828: 15.623736029803087
1829: 18.914679385267622
1830: 19.94102431453699
1831: 20.252832131822863
1832: 18.167397020157757
1833: 22.844265439202147
1834: 17.816445830085737
1835: 19.74567806377257
1836: 22.259059122695486
1837: 20.14407894736842
1838: 21.130470236477304
1839: 23.25336838646073
1840: 19.12202807

In [1]:
year = 1800 
avg_precip = 0 
while(avg_precip < 22):
    import pandas as pd
    from pathlib import Path
    myDF = []
    myDF = pd.read_csv(Path(f'/anvil/projects/tdm/data/noaa/{year}.csv'), header=None, names=["id","date","element_code","value","mflag","qflag","sflag","obstime"]) 
    prcp_data = myDF[myDF['element_code'] == 'PRCP']
    avg_precip = prcp_data['value'].mean() 
    print(f"{year}: {avg_precip}")
    year = year + 1 

1800: 16.016438356164382
1801: 19.416438356164385
1802: 11.37945205479452
1803: 19.63013698630137
1804: 16.968903436988544
1805: 13.963013698630137
1806: 13.481361426256077
1807: 15.396825396825397
1808: 16.23633879781421
1809: 14.687671232876712
1810: 14.23489010989011
1811: 12.754794520547945
1812: 15.904371584699453
1813: 22.049397590361444


For part a, I took a TA's help. His name is Sharan Sivakumar. He gave me the code structure for parts: 
prcp_data = myDF[myDF['element_code'] == 'PRCP'] 
avg_precip = prcp_data['value'].mean() 
print(f"{year}: {avg_precip}") 

The code above is my modified version. But he gave me the general idea. So, in his given structure, I input my values such as "myDF", and "element_code", "PRCP". But, the only problem was that I did not know how to input the mean() function. Another problem which he caught was that my data was being read as a list instead of a dataframe. So instead of just inputting the dataset, he adviced me to input it as: Path(f'/anvil/projects/tdm/data/noaa/{year}.csv') so that my dataset is read like a dataframe. This made it easier to pull out the object: element_code because previously, I was getting an error everytime I tried to modify the code. 

For part b, I got the code structure from Dr. Ward in Piazza. He gave the following structure: 

year = 1800

avg_precip = 0

while(avg_precip < 22)

   calculate the avg_precip for the current year

   year = year + 1
   
I later, modified this code structure and inputed my values; formula and data for the average value calculations, so as to meet the question's requirements. 

## Question 3

In [31]:
import pandas as pd
myDF = pd.read_csv("/anvil/projects/tdm/data/noaa/1880.csv" , header=None , names=["id","date","element_code","value","mflag","qflag","sflag","obstime"])
myDF[myDF['element_code'] == 'PRCP'].groupby('id')['value'].mean().sort_values() 

id
USC00308138      0.000000
ASN00033054      0.000000
ASN00033050      0.000000
ASN00033046      0.000000
ASN00033007      0.000000
                  ...    
USC00314055     88.978632
USC00316001    103.685039
USC00315838    112.875000
USC00012207    151.290698
USC00483073    610.000000
Name: value, Length: 744, dtype: float64

In [46]:
import pandas as pd
myDF = pd.read_csv("/anvil/projects/tdm/data/noaa/1880.csv" , header=None , names=["id","date","element_code","value","mflag","qflag","sflag","obstime"])
myDF[myDF['element_code'] == 'PRCP'].groupby('id')['value'].mean().sort_values()['USC00288878'] 

35.08516483516483

For part and b, I got the code structure by watching the video, "extracting values from a Python series and turning a series into a dictionary" linked under the videos for project 4: https://the-examples-book.com/programming-languages/python/some-examples-for-TDM-10200-project-4. But I replaced the specific "id" value to 'USC00288878' so that I can meet the question's requirements. 

For part a, USC00483073 is the 'id' that has the largest average precipitation. 
For part b, the average precipiation for 'id' USC00288878 = 35.08516483516483. 

## Question 4

In [47]:
import pandas as pd
myDF = pd.read_csv("/anvil/projects/tdm/data/noaa/1880.csv" , header=None , names=["id","date","element_code","value","mflag","qflag","sflag","obstime"])
myDF[myDF['element_code'] == 'PRCP'].groupby('id')['value'].mean().sort_values().to_dict() 

{'USC00308138': 0.0,
 'ASN00033054': 0.0,
 'ASN00033050': 0.0,
 'ASN00033046': 0.0,
 'ASN00033007': 0.0,
 'ASN00033005': 0.0,
 'ASN00032047': 0.0,
 'ASN00032004': 0.0,
 'ASN00030036': 0.0,
 'ASN00030018': 0.0,
 'ASN00029072': 0.0,
 'ASN00029041': 0.0,
 'ASN00033065': 0.0,
 'ASN00029019': 0.0,
 'ASN00043030': 0.0,
 'ASN00026016': 0.0,
 'ASN00044004': 0.0,
 'ASN00024573': 0.0,
 'ASN00024531': 0.0,
 'ASN00023795': 0.0,
 'ASN00023782': 0.0,
 'ASN00023773': 0.0,
 'ASN00023068': 0.0,
 'ASN00044026': 0.0,
 'ASN00045014': 0.0,
 'ASN00043022': 0.0,
 'ASN00045017': 0.0,
 'ASN00035019': 0.0,
 'ASN00035049': 0.0,
 'ASN00041038': 0.0,
 'ASN00041023': 0.0,
 'ASN00041013': 0.0,
 'ASN00040358': 0.0,
 'ASN00040242': 0.0,
 'ASN00040226': 0.0,
 'ASN00040225': 0.0,
 'ASN00042071': 0.0,
 'ASN00040126': 0.0,
 'ASN00040111': 0.0,
 'ASN00040101': 0.0,
 'ASN00035037': 0.0,
 'ASN00040096': 0.0,
 'ASN00040047': 0.0,
 'ASN00040016': 0.0,
 'ASN00039041': 0.0,
 'ASN00039039': 0.0,
 'ASN00039022': 0.0,
 'ASN00039003

For question 4, part a, I used the same code structure that I used for question 3. However, I also added the "".to_dict()" fuction to covert my information from a series to a dictionary. Hence, I got the code structure by watching the video, "extracting values from a Python series and turning a series into a dictionary" linked under the videos for project 4: https://the-examples-book.com/programming-languages/python/some-examples-for-TDM-10200-project-4. 

## Question 5

In [11]:
import pandas as pd
myDF = pd.read_csv("/anvil/projects/tdm/data/flights/subset/airports.csv")
print(myDF) 

     iata                    airport              city state country  \
0     00M                   Thigpen        Bay Springs    MS     USA   
1     00R       Livingston Municipal        Livingston    TX     USA   
2     00V                Meadow Lake  Colorado Springs    CO     USA   
3     01G               Perry-Warsaw             Perry    NY     USA   
4     01J           Hilliard Airpark          Hilliard    FL     USA   
...   ...                        ...               ...   ...     ...   
3371  ZEF            Elkin Municipal             Elkin    NC     USA   
3372  ZER  Schuylkill Cty/Joe Zerbey        Pottsville    PA     USA   
3373  ZPH      Zephyrhills Municipal       Zephyrhills    FL     USA   
3374  ZUN                 Black Rock              Zuni    NM     USA   
3375  ZZV       Zanesville Municipal        Zanesville    OH     USA   

            lat        long  
0     31.953765  -89.234505  
1     30.685861  -95.017928  
2     38.945749 -104.569893  
3     42.741347

In [12]:
import pandas as pd
myDF = pd.read_csv("/anvil/projects/tdm/data/flights/subset/airports.csv")

for index, row in myDF.iterrows(): 
    if row["city"] == "Colorado Springs":
        print(row["airport"]) 

Meadow Lake
City of Colorado Springs Muni


In [19]:
import pandas as pd
mydataset = pd.read_csv("http://data.insideairbnb.com/belgium/vlg/antwerp/2023-06-28/visualisations/listings.csv") 
print(mydataset) 

                      id                                               name  \
0                  50904  Boutique hotel in Antwerp · ★5.0 · 1 bedroom ·...   
1                 224682  Rental unit in Antwerpen · ★4.78 · 1 bedroom ·...   
2                 345959  Rental unit in Antwerp · ★4.80 · 2 bedrooms · ...   
3                 363155  Rental unit in Antwerp · ★4.85 · 1 bedroom · 1...   
4                 366252  Rental unit in Antwerp · ★4.62 · 1 bedroom · 1...   
...                  ...                                                ...   
2535  922495587108700765  Rental unit in Antwerpen · ★New · 1 bedroom · ...   
2536  922853557828818693  Rental unit in Antwerpen · ★New · 1 bedroom · ...   
2537  922992228688558510  Rental unit in Antwerpen · ★New · 1 bedroom · ...   
2538  923028396839396315  Rental unit in Antwerpen · ★New · 1 bedroom · ...   
2539  923135124226357523  Rental unit in Antwerpen · ★New · 1 bedroom · ...   

        host_id   host_name  neighbourhood_group  \

In [26]:
import pandas as pd
mydataset = pd.read_csv("http://data.insideairbnb.com/belgium/vlg/antwerp/2023-06-28/visualisations/listings.csv") 

for index, row in mydataset.iterrows(): 
    if row["id"] == 50904:
        print(row["room_type"])  

Hotel room


For both the datasets, I used a 'for loop' because I was more comfortable using this type of loop structure. I selected both the datasets from the Projects archive in the examples textbook: https://the-examples-book.com/projects/data-sets/AirBnB and https://the-examples-book.com/projects/data-sets/flights. When I used 'for loop', I got the code structure from question 1 of project 4. However, I modified that code to extract information relevant to my datasets. This gave me additional practice when working with 'for loop' as I struggled with incorporating them for the second question.  

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.